gluonnlp/model/utils.py

```diff
@@ -207,3 +209,36 @@ def _get_rnn_layer(mode, num_layers, input_size, hidden_size, dropout, weight_dr
         apply_weight_drop(block, '.*h2h_weight', rate=weight_dropout)
 
     return block
+
+
+def _multi_gpu_clip_global_norm_scale(arrays, max_norm):
+    """Compute the global norm'scale in order to make the 2-norm smaller than `max_norm`.
+    """
+    assert len(arrays) > 0
+    ctx = arrays[0].context
+    total_norm = ndarray.add_n(*[ndarray.dot(x, x).as_in_context(ctx)
+                                 for x in (arr.reshape((-1,)) for arr in arrays)])
+    total_norm = ndarray.sqrt(total_norm).asscalar()
+    if not np.isfinite(total_norm):
+        warnings.warn(UserWarning('nan or inf is detected. Clipping results will be undefined.'),
+                      stacklevel=2)
+    scale = max_norm / (total_norm + 1e-8)
+    return total_norm, scale
+
+
+def multi_gpu_clip_global_norm(trainer, parameters, max_norm):
+    """Rescales NDArrays so that the sum of their 2-norm on every context
+    is smaller than `max_norm`.
+    """
+    trainer.allreduce_grads()
+    ctx = list(parameters)[0].list_ctx()[0]
+    global_grads = [p.grad(ctx) for p in parameters]
+    total_norm, scale = _multi_gpu_clip_global_norm_scale(global_grads, max_norm)
```

Comment:
`trainer.allreduce_grads()` does not perform averaging, so this should be taken into account when computing the scale.

Reply:
Since now the experiment is running with single gpu, it should be fine. I can further test it after KDD.

Reply:
Since the script is not used in KDD, let's get it right and not rush this.